### PR TITLE
dns: add dns proxy stats

### PR DESCRIPTION
**Pull request by ramaraochavali**
Fixes https://github.com/istio/istio/issues/33222
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/dns/monitoring.go
Comment (howardjohn) [inline pilot/pkg/dns/monitoring.go:30]: I think prometheus standard (and openmetrics requirement) is that it ends with _total. So `dns_requests_total`.

File: pilot/pkg/dns/monitoring.go
Comment (howardjohn) [inline pilot/pkg/dns/monitoring.go:35]: @douglas-reid should this be a tag?

File: pilot/pkg/dns/monitoring.go
Comment (howardjohn) [inline pilot/pkg/dns/monitoring.go:45]: cardinality may be an issue here @nrjpoddar @douglas-reid any guidance?

**Review by howardjohn — COMMENTED** [review]
I think the idea here is good, but we should have careful review from telemetry folks @istio/wg-policies-and-telemetry-maintainers

**Comment by ramaraochavali** [comment]
@mandarjog @douglas-reid can you PTAL?

**Comment by ramaraochavali** [comment]
@douglas-reid can you PTAL?

File: pilot/pkg/dns/monitoring.go
Comment (nrjpoddar) [reply pilot/pkg/dns/monitoring.go:45]: Are these metrics going to be enabled by default when DNS proxy will be enabled or is it going to be opt in?

Comment (ramaraochavali) [reply pilot/pkg/dns/monitoring.go:45]: It is not opt-in currently. Enabled when dns proxy is on.

**Comment by ramaraochavali** [comment]
@howardjohn @nrjpoddar @douglas-reid gentle ping. Can you PTAL when you get chance?

File: pilot/pkg/dns/monitoring.go
Comment (nrjpoddar) [reply pilot/pkg/dns/monitoring.go:45]: So if a user has enabled DNS_PROXY, then on upgrades they will see an increase in their metric storage after this change. Should we guard it and keep default off?

Comment (ramaraochavali) [reply pilot/pkg/dns/monitoring.go:45]: yes. But since DNS proxy is still experimental and the amount of metrics we are adding here are not too many - I am guessing it should be fine. However, I am not against adding a flag defaulted to false, if you think it is better

Comment (nrjpoddar) [reply pilot/pkg/dns/monitoring.go:45]: I will let others weigh in too but an opt-in option might be better

Comment (mandarjog) [reply pilot/pkg/dns/monitoring.go:45]: adding 8 bucket distribution is like adding another istio standard distribution metric sans other labels. We can probably reduce it to 6. So I don’t think cardinality is a big issue. Since dns itself is experimental we can leave this in and create an option based on feedback.

we many need to systematically add metrics opt in for everything for the internal metric generation, but does not have to be here on an adhoc basis.

Comment (ramaraochavali) [reply pilot/pkg/dns/monitoring.go:45]: Thanks I have reduced it to 6 buckets. @howardjohn can you PTAL?

Comment (nrjpoddar) [reply pilot/pkg/dns/monitoring.go:45]: SGTM.

**Review by nrjpoddar — APPROVED** [review]

File: pilot/pkg/dns/monitoring.go
Comment (hzxuzhonghu) [inline pilot/pkg/dns/monitoring.go:25]: Just one point, why register these metrics when even dns proxy is unused.

Comment (hzxuzhonghu) [reply pilot/pkg/dns/monitoring.go:25]: Hha, i am not a fan of doing everything in `init`

Comment (ramaraochavali) [reply pilot/pkg/dns/monitoring.go]: Fixed

**Review by hzxuzhonghu — COMMENTED** [review]
lgtm

File: pilot/pkg/dns/monitoring.go
Comment (howardjohn) [reply pilot/pkg/dns/monitoring.go:38]: Should we use seconds instead of ms? See https://prometheus.io/docs/practices/naming/#base-units

File: pilot/pkg/dns/monitoring.go
Comment (howardjohn) [inline pilot/pkg/dns/monitoring.go:33]: dns_total_upstream_failures -> dns_upstream_failures_total to follow https://prometheus.io/docs/practices/naming/#metric-names (and consistency with upstreamRequests)

**Pull request merged**
